api/client/run.go

```diff
@@ -62,6 +72,179 @@ func runStartContainerErr(err error) error {
+		(count >= 2 && strings.HasSuffix(part[count-2], ".git")):
+		return "git"
+	case strings.HasPrefix(source, "http://"):
+		fallthrough
+	case strings.HasPrefix(source, "https://"):
+		return "http"
+	case strings.HasPrefix(source, "/"):
+		return "local"
+	default:
+		return "unknown"
+	}
+}
+
+func (cli *DockerCli) initSpecialVolumes(config *container.Config, hostConfig *container.HostConfig, networkingConfig *networktypes.NetworkingConfig, initvols []*InitVolume) error {
+	const INIT_VOLUME_PATH = "/vol/"
+	const INIT_VOLUME_IMAGE = "hyperhq/volume_uploader:latest"
+	var (
+		initConfig     *container.Config
+		initHostConfig *container.HostConfig
+		errCh          chan error
+		execCount      uint32
+		fip            string
+	)
+
+	initConfig = &container.Config{
+		User:       config.User,
+		Image:      INIT_VOLUME_IMAGE,
+		StopSignal: config.StopSignal,
+	}
+
+	initHostConfig = &container.HostConfig{
+		Binds:      make([]string, 0),
+		DNS:        hostConfig.DNS,
+		DNSOptions: hostConfig.DNSOptions,
+		DNSSearch:  hostConfig.DNSSearch,
+		ExtraHosts: hostConfig.ExtraHosts,
+	}
+
+	for _, vol := range initvols {
+		initHostConfig.Binds = append(initHostConfig.Binds, vol.Name+":"+INIT_VOLUME_PATH+vol.Destination)
+	}
+	passwd := uuid.NewV1()
+	initConfig.Env = append(config.Env, "ROOTPASSWORD="+passwd.String(), "LOCALROOT="+INIT_VOLUME_PATH)
+
+	createResponse, err := cli.createContainer(initConfig, initHostConfig, networkingConfig, hostConfig.ContainerIDFile, "")
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			if _, rmErr := cli.removeContainer(createResponse.ID, true, false, true); rmErr != nil {
+				fmt.Fprintf(cli.err, "clean up init container failed: %s\n", rmErr.Error())
+			}
+		}
+		if fip != "" {
+			if rmErr := cli.releaseFip(fip); rmErr != nil {
+				fmt.Fprintf(cli.err, "failed to clean up container fip %s: %s\n", fip, rmErr.Error())
+			}
+		}
+	}()
+
+	if err = cli.client.ContainerStart(createResponse.ID); err != nil {
+		return err
+	}
+
+	errCh = make(chan error, len(initvols))
+	for _, vol := range initvols {
+		var cmd []string
+		volType := checkSourceType(vol.Source)
+		switch volType {
+		case "git":
+			cmd = append(cmd, "git", "clone", vol.Source, INIT_VOLUME_PATH+vol.Destination)
+		case "http":
+			parts := strings.Split(vol.Source, "/")
+			cmd = append(cmd, "wget", "--no-check-certificate", "--tries=5", "--mirror", "--no-host-directories", "--cut-dirs="+strconv.Itoa(len(parts)), vol.Source, "--directory-prefix="+INIT_VOLUME_PATH+vol.Destination)
+		case "local":
+			execCount++
+			if fip == "" {
+				fip, err = cli.associateNewFip(createResponse.ID)
+				if err != nil {
+					return err
+				}
+			}
+			go func() {
+				err := cli.uploadLocalResource(vol.Source, INIT_VOLUME_PATH+vol.Destination, fip, "root", passwd.String())
+				if err != nil {
+					err = fmt.Errorf("Failed to upload %s: %s", vol.Source, err.Error())
+				}
+				errCh <- err
```

Comment:
There is a bug here. the var `vol` is an encapsulated variable. the two threads share the `vol`, but the loop thread(`for _, vol := range initvols {`) may change the `vol` before this uploader thread, and the uploader thread use the wrong `vol`.

My test hits this bug
```
./hyper/hyper run -t -i -v https://github.com/nginxinc/docker-nginx/raw/master/stable/alpine/Dockerfile:/data1 -v https://github.com/nginxinc/docker-nginx.git:/data2 -v /Users/laijs/work/gopath/src/github.com/hyperhq/hypercli/hyper/:/data3 -v /Users/laijs/work/gopath/src/github.com/hyperhq/hypercli/man/:/data4 busybox sh
./hyper/hyper: Failed to upload https://github.com/nginxinc/docker-nginx/raw/master/stable/alpine/Dockerfile: stat https://github.com/nginxinc/docker-nginx/raw/master/stable/alpine/Dockerfile: no such file or directory.
```

simple fix
```patch
diff --git a/api/client/run.go b/api/client/run.go
index 30a66df..b21f0c4 100644
--- a/api/client/run.go
+++ b/api/client/run.go
@@ -186,13 +186,13 @@ func (cli *DockerCli) initSpecialVolumes(config *container.Config, hostConfig *c
 					return err
 				}
 			}
-			go func() {
+			go func(vol *InitVolume) {
 				err := cli.uploadLocalResource(vol.Source, INIT_VOLUME_PATH+vol.Destination, fip, "root", passwd.String())
 				if err != nil {
 					err = fmt.Errorf("Failed to upload %s: %s", vol.Source, err.Error())
 				}
 				errCh <- err
-			}()
+			}(vol)
 		default:
 			continue
 		}
```